File: apps/worker/src/processors/score.ts
import type { Job } from 'bullmq'
import { getDb, getJobById, updateJob, logEvent } from '@jobpilot/db'
import { getProvider, extractJson } from '@jobpilot/ai'
import { MODEL_FAST } from '@jobpilot/config'
import { loadProfile } from '@jobpilot/db'

export async function processScore(job: Job<{ jobId: number }>): Promise<{ jobIds: number[] }> {
  const db = getDb()
  const dbJob = getJobById(db, job.data.jobId)
  if (!dbJob) return { jobIds: [] }

  const desc = (dbJob.description || '').slice(0, 1500)
  if (!desc) {
    updateJob(db, dbJob.id, { status: 'skipped', score: 0, scoreReason: 'Sem descrição' })
    return { jobIds: [] }
  }

  const ai = getProvider()
  const profileText = loadProfile()
  const candidateInfo = profileText.slice(0, 800) || 'Nenhum perfil configurado'
  const threshold = 40

  const prompt =
    `Score 0-100 job-candidate match. Reply ONLY with JSON.\n\n` +
    `Job: ${dbJob.title} at ${dbJob.company} (${dbJob.location || 'N/A'})\n` +
    `Description: ${desc}\n\n` +
    `Candidate:\n${candidateInfo}\n\n` +
    `Reply ONLY: {"score": N, "reason": "..."}`

  try {
    const raw = await ai.complete(prompt, { model: MODEL_FAST })
    const result = extractJson(raw)
    const score = Number(result.score)
    const reason = String(result.reason || '')

    if (score < threshold) {
      updateJob(db, dbJob.id, { score, scoreReason: reason, status: 'skipped' })
      logEvent(db, dbJob.id, 'score', 'info', `Pontuação ${score} < ${threshold}, ignorada`)
      return { jobIds: [] }
    }

    updateJob(db, dbJob.id, { score, scoreReason: reason, status: 'scored' })
    logEvent(db, dbJob.id, 'score', 'info', `Pontuação ${score}: ${reason}`)
    return { jobIds: [dbJob.id] }
  } catch (err) {
    logEvent(db, dbJob.id, 'score', 'error', `Erro na pontuação: ${err}`)
    updateJob(db, dbJob.id, { status: 'skipped', score: 0, scoreReason: `Erro: ${err}` })
    return { jobIds: [] }
  }
}

File: apps/agent/pipeline/score.py
"""Stage 3: Score jobs against candidate profile using AI."""

from __future__ import annotations

import json
import logging
import sqlite3

from apps.backend.ai_provider import extract_json, get_provider
from apps.agent.config import MODEL_FAST
from apps.agent.db import get_jobs, update_job, log_event

logger = logging.getLogger(__name__)


def run(db: sqlite3.Connection, profile: dict, threshold: int = 40) -> int:
    """Score enriched jobs. Returns count of scored jobs."""
    jobs = get_jobs(db, status="enriched")
    scored = 0
    ai = get_provider()

    # Build a short candidate summary once
    cv_short = profile.get("base_cv", "")[:800]
    candidate = f"Name: {profile.get('name','N/A')}, Location: {profile.get('location','N/A')}\n{cv_short}"

    for job in jobs:
        desc = (job["description"] or "")[:1500]
        if not desc:
            update_job(db, job["id"], status="skipped", score=0, score_reason="No description")
            continue

        prompt = (
            f"Score 0-100 job-candidate match. Reply ONLY with JSON.\n\n"
            f"Job: {job['title']} at {job['company']} ({job['location'] or 'N/A'})\n"
            f"Description: {desc}\n\n"
            f"Candidate: {candidate}\n\n"
            f'Reply ONLY: {{"score": N, "reason": "..."}}'
        )

        try:
            raw = ai.complete(prompt, model=MODEL_FAST)
            result = extract_json(raw)
            score = int(result["score"])
            reason = result.get("reason", "")

            if score < threshold:
                update_job(db, job["id"], score=score, score_reason=reason, status="skipped")
                log_event(db, job["id"], "score", "info", f"Score {score} < {threshold}, skipped")
            else:
                update_job(db, job["id"], score=score, score_reason=reason, status="scored")
                log_event(db, job["id"], "score", "info", f"Score {score}: {reason}")
                scored += 1

        except (json.JSONDecodeError, KeyError, ValueError) as e:
            logger.warning("Score parse fail job %d: %s | raw=%s", job["id"], e, repr(raw[:120]) if 'raw' in dir() else '?')
            log_event(db, job["id"], "score", "error", f"Parse error: {e}")
            update_job(db, job["id"], status="skipped", score=0, score_reason=f"Parse error: {e}")
        except Exception as e:
            logger.error("AI scoring failed for job %d: %s", job["id"], e)
            log_event(db, job["id"], "score", "error", f"AI error: {e}")

    return scored
